### PR TITLE
[DomCrawler] Added Crawler::matches(), ::closest(), ::outerHtml()

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #18609
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/12225

### CODE REVIEW

File: src/Symfony/Component/DomCrawler/Crawler.php [hunk 1]
```diff
@@ -427,6 +427,45 @@ public function siblings()
...
+        }
+
+        $converter = $this->createCssSelectorConverter();
+        $xpath = $converter->toXPath($selector, 'self::');
+
+        return 0 !== $this->filterRelativeXPath($xpath)->count();
+    }
+
+    /**
+     * Return first parents (heading toward the document root) of the Element that that matches the provided selector.
```
Comment (fabpot) [inline src/Symfony/Component/DomCrawler/Crawler.php]: `Returns the`

double `that`

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/DomCrawler/Crawler.php [hunk 2]
```diff
@@ -427,6 +427,45 @@ public function siblings()
         return $this->createSubCrawler($this->sibling($this->getNode(0)->parentNode->firstChild));
     }
```
[the remark this answers is not in the document]
Comment (lyrixx) [reply src/Symfony/Component/DomCrawler/Crawler.php]: Actually I was lazy. And I think it was so obvious that it is not needed

File: src/Symfony/Component/DomCrawler/Crawler.php (hunk 1 above)
Comment (lyrixx) [reply src/Symfony/Component/DomCrawler/Crawler.php]: Thanks. Fixed

Comment (fabpot) [reply src/Symfony/Component/DomCrawler/Crawler.php]: The `that that` does not seem to be fixed.

File: src/Symfony/Component/DomCrawler/Crawler.php [hunk 3]
```diff
@@ -427,6 +427,45 @@ public function siblings()
...
 
+    public function matches(string $selector): bool
+    {
+        if (!$this->nodes) {
+            throw new \InvalidArgumentException('The current node list is empty.');
+        }
```
Comment (fabpot) [inline src/Symfony/Component/DomCrawler/Crawler.php]: Is it needed, in other similar cases, we could return `false` here.

**Review by fabpot — APPROVED** [review]
Left 2 minor comments

File: src/Symfony/Component/DomCrawler/Crawler.php (hunk 3 above)
Comment (lyrixx) [reply src/Symfony/Component/DomCrawler/Crawler.php]: I'm not sure to understand but I updated the code to return false

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @lyrixx.

**Pull request merged**